### PR TITLE
Implement getters and setters for EntityItem owner and thrower

### LINKED ISSUE #1526: Item Thrower and Owner API.
Something that would be useful, and never being implemented (Not sure why though), is an API for Item to getting/setting the Thrower and the Owner of an Item. 
When the Thrower is almost useless but good for completition, Owner is used by vanilla to avoid other players to pick up your item, and since it is already done I don't see why I should do it with events.

What I am asking here is a getter and a setter for both UUID (Owner and Thrower) saved inside the EntityItem class. I would do it by myself, but I haven't got much time, unfortunately.

### PR DESCRIPTION
**Pull request by BillyGalbreath**
Closes #1526

### CODE REVIEW

File: Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+@@ -21,6 +23,13 @@ public class CraftItem extends CraftEntity implements Item {
+         this(server, entity, entity);
+     }
+ 
++    // Paper start
++    @Override
++    public EntityItem getHandle() {
++        return (EntityItem) entity;
++    }
++    // PAper end
```
Comment (kashike) [inline Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch]: :no_good_woman:

File: Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ public class CraftItem extends CraftEntity implements Item {
+     private final EntityItem item;
+ 
+@@ -21,6 +23,13 @@ public class CraftItem extends CraftEntity implements Item {
+         this(server, entity, entity);
+     }
+ 
++    // Paper start
++    @Override
++    public EntityItem getHandle() {
```
Comment (aikar) [inline Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch:61]: Hmm, This should still be safe for any plugins using reflection right?
@DemonWav @kashike ?

**Review by aikar — APPROVED** [review]

File: Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch (hunk 2 above)
Comment (BillyGalbreath) [reply Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch:61]: If not I could change it to use the `item` field stored there. Honestly didnt even notice it until just now.

File: Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+index f596b4351..f21a5915a 100644
+--- a/src/main/java/net/minecraft/server/EntityItem.java
++++ b/src/main/java/net/minecraft/server/EntityItem.java
+@@ -410,20 +410,24 @@ public class EntityItem extends Entity {
+         this.getDataWatcher().set(EntityItem.b, itemstack);
+     }
+ 
++    public UUID getOwner() { return k(); } // Paper - OBFHELPER
+     @Nullable
+     public UUID k() {
```
Comment (Black-Hole) [inline Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch]: Maybe a Paper comment on this line to force a conflict if the method name is changed?

File: Spigot-API-Patches/0159-Implement-getters-and-setters-for-EntityItem-owner-a.patch [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+diff --git a/src/main/java/org/bukkit/entity/Item.java b/src/main/java/org/bukkit/entity/Item.java
+index 2e98e4f9..439cf3ab 100644
+--- a/src/main/java/org/bukkit/entity/Item.java
++++ b/src/main/java/org/bukkit/entity/Item.java
+@@ -2,6 +2,8 @@ package org.bukkit.entity;
+ 
+ import org.bukkit.inventory.ItemStack;
+ 
++import java.util.UUID;
```
Comment (electronicboy) [inline Spigot-API-Patches/0159-Implement-getters-and-setters-for-EntityItem-owner-a.patch:18]: drop or comment

File: Spigot-API-Patches/0159-Implement-getters-and-setters-for-EntityItem-owner-a.patch [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
++
+ /**
+  * Represents an Item.
+  */
+@@ -49,5 +51,33 @@ public interface Item extends Entity {
+      * @param canMobPickup True to allow non-player entity pickup
+      */
+     public void setCanMobPickup(boolean canMobPickup);
++
++    /**
```
Comment (electronicboy) [inline Spigot-API-Patches/0159-Implement-getters-and-setters-for-EntityItem-owner-a.patch:29]: nullability annotations are always nice! 😉

File: Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch (hunk 3 above)
Comment (electronicboy) [reply Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch]: This should conflict as-is, so non-issue here :)

File: Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch (hunk 2 above)
Comment (electronicboy) [reply Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch:61]: This should be fine, and CB defines getHandle methods like that all over the place without any issues

**Review by electronicboy — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0159-Implement-getters-and-setters-for-EntityItem-owner-a.patch (hunk 4 above)
Comment (BillyGalbreath) [reply Spigot-API-Patches/0159-Implement-getters-and-setters-for-EntityItem-owner-a.patch:18]: was told by @aikar not to comment imports

Comment (aikar) [reply Spigot-API-Patches/0159-Implement-getters-and-setters-for-EntityItem-owner-a.patch:18]: I said I generally don't care about them. I wouldn't 'fail' a review over it.

File: Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch (hunk 3 above)
Comment (aikar) [reply Spigot-Server-Patches/0386-Implement-getters-and-setters-for-EntityItem-owner-a.patch]: I think I have seen apatch be smart enough to auto apply this without conflict due to the annotation in middle... but we have a few like this.
ultimately the simple solution is to put the obf helper before the nullable itself is what I've done, so the k's anno is at end of the patched line (before the paper comment)

that's also why we keep the OBFHELPER consistent, so we can do a search of the code base and find wrong ones too.

**Review by electronicboy — APPROVED** [review]

**Review by zachbr — APPROVED** [review]

**Comment by zachbr** [comment]
Rebased and merged manually with 52ae2a1aec33089ad4704e029e960c3df7168ed8
GitHub and I are still having some sort of disagreement about being able to push to requesters repos

img

Thanks

**Pull request closed without merging**